c/src/lib/libbsp/arm/shared/stm32fxxxx/can/hal-can.c

```diff
@@ -0,0 +1,664 @@
+#define CAN2_GPIO_PORT                 GPIOB
+#define CAN2_AF                        GPIO_AF9_CAN2
+
+#define MAX_FILTERS                   14  
+
+CAN_Status stm32_can_start_tx
+(
+  CAN_HandleTypeDef* hCanHandle, 
+  can_msg * msg
+)
+{
+  //TODO refactor this to not have so much STM32 stuff
+  CAN_Status Status;
+
+  hCanHandle->pTxMsg->RTR     = CAN_RTR_DATA;
+  hCanHandle->pTxMsg->IDE     = CAN_ID_STD;
+
+  hCanHandle->pTxMsg->StdId   = msg->id;
+  hCanHandle->pTxMsg->DLC     = msg->len;
+  memcpy(hCanHandle->pTxMsg->Data, msg->data, 8);
+
+  Status = HAL_CAN_Transmit_IT(hCanHandle);
+  return Status;
+}
+
+
+CAN_Status stm32_can_start_rx
+(
+  CAN_HandleTypeDef *hCanHandle
+)
+{
+  //TODO use both FIFOs
+  CAN_Status Status;
+  Status = HAL_CAN_Receive_IT(hCanHandle, 0);
+  return Status;
+}
+
+static int local_roundf
+(
+  float arg
+)
+{
+  int ret = (int) arg;
+
+  if(arg > 0.0F) {
+    if((arg - ret) >= 0.5) ret++;
+  } else {
+    if((arg - ret) <= -0.5) ret--;
+  }
+
+  return ret;
+}
+
+void HAL_CAN_RxCpltCallback
+(
+  CAN_HandleTypeDef *hCanHandle
+)
+{
+  // Wakeup RX Task
+  rtems_status_code sc;
+  stm32_can_bus * bus = (((HandleWrapper *) hCanHandle)->bus);
+  volatile int a;
+  sc = rtems_event_transient_send(bus->base.rx_task_id);
+  a++;
+}
+
+
+rtems_task stm32_rx_task
+(
+  rtems_task_argument arg
+)
+{
+  rtems_status_code sc;
+  can_msg msg;
+  size_t len;
+
+  stm32_can_bus * bus = (stm32_can_bus *) arg;
+  CAN_HandleTypeDef * hCanHandle = (CAN_HandleTypeDef * ) &bus->wrapper;
+
+  while (1) {
+    // Schedule a read
+    stm32_can_start_rx(hCanHandle);
+
+
+
+    // Wait for interrupt
+    sc = rtems_event_transient_receive(RTEMS_WAIT, RTEMS_NO_TIMEOUT);
+    _Assert(sc == RTEMS_SUCCESSFUL);
+
+    //TODO consider 29 bit ids
+    msg.id = hCanHandle->pRxMsg->StdId;
+    msg.len = hCanHandle->pRxMsg->DLC;
+    memcpy(&msg.data, &hCanHandle->pRxMsg->Data, msg.len);
+    // Pass message along to stack
+    rtems_message_queue_send(bus->base.rx_msg_queue, &msg, sizeof(msg));
+  }
+}
+
+
+rtems_task stm32_tx_task
+(
+  rtems_task_argument arg
+)
+{
+  rtems_status_code sc;
+  can_msg msg;
+  size_t len;
+
+  stm32_can_bus * bus = (stm32_can_bus *) arg;
+  CAN_HandleTypeDef * hCanHandle = &(bus->wrapper.handle);
+  while (1) {
+    //TODO abstract away into
+    // can_msg_queue_receive
+    sc = rtems_message_queue_receive( 
+        bus->base.tx_msg_queue, 
+        &msg,
+        &len,
+        RTEMS_WAIT,
+        RTEMS_NO_TIMEOUT
+      );
+    _Assert(len == sizeof(msg));
+
+    stm32_can_start_tx(hCanHandle, &msg);
+  }
+}
+
+
+void stm32_can_gpio_init
+(
+  CAN_Instance canInstance
+)
+{
+  GPIO_InitTypeDef GPIO_InitStructure;
+
+  // Enable the clock CAN Core //
+  __HAL_RCC_CAN1_CLK_ENABLE();
+
+  if (canInstance == CAN_TWO)
+  {
+    __HAL_RCC_CAN2_CLK_ENABLE();	// CAN2 requires clock for CAN1 enabled too
+  }
+
+  // CAN GPIO Clock Enable//
+  __GPIOB_CLK_ENABLE();
+
```

Comment:
The GPIO port clock has to be enabled depending on the port and pins used. Also depending on which instance (if different instance uses different GPIO ports)